pkg/expression/contextopt/optional.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package contextopt
+
+import (
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/pkg/expression/context"
+	"github.com/pingcap/tidb/pkg/util/intest"
+)
+
+// RequireOptionalEvalProps is the interface for the function that requires optional evaluation properties or not.
+type RequireOptionalEvalProps interface {
+	// RequiredOptionalEvalProps returns the optional properties that this function requires.
+	// If the returned `OptionalEvalPropKeySet` is empty,
+	// it means this function does not require any optional properties.
+	RequiredOptionalEvalProps() context.OptionalEvalPropKeySet
```

Comment:
This function is not used in this PR. Can you describe the usage scenario of this function?

Reply:
The return value of this function has serveral usages in some future works:

- To determine whether the provided `EvalContext` can be used by some expressions by checking whether the required properties is provided by `EvalContext`
- For cursor fetch, when we need to "clone" a `EvalContext`, we can check whether all properties required by expressions in executors are able to be cloned. If not, we cannot clone the `EvalContext`, otherwise, the context can be cloned safely by only cloning the required properties